### PR TITLE
ADBDEV-3674-new: new shouldn't return NULL

### PR DESCRIPTION
**Pull request by RekGRpth**
ORCA makes extensive use of the [GPOS_NEW](https://github.com/arenadata/gpdb/blob/88532d712d0258ab14e5992854bd04605d1e763f/src/backend/gporca/libgpos/include/gpos/memory/CMemoryPool.h#LL299C9-L299C17) to allocate memory, not checking the result for NULL, but instead relying on throwing an exception.
[The implementation of this macro](https://github.com/arenadata/gpdb/blob/88532d712d0258ab14e5992854bd04605d1e763f/src/backend/gporca/libgpos/include/gpos/memory/CMemoryPool.h#L276-L281) depends on the memory pool.
Depending on the variable [optimizer_use_gpdb_allocators](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gpopt/CGPOptimizer.cpp#L174-L177) can be used [either](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gpopt/utils/CMemoryPoolPallocManager.cpp#L53-L58) GreenPlum's memory pool manager [CMemoryPoolPalloc](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gpopt/utils/CMemoryPoolPalloc.cpp#L33-L61) (default), [or](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/memory/CMemoryPoolManager.cpp#L69-L79) ORCA's memory pool manager [CMemoryPoolTracker](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/memory/CMemoryPoolTracker.cpp#L67-L118).
When there is not enough memory for the GreenPlum's memory pool manager in the [gpdb::GPDBMemoryContextAlloc](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gpopt/gpdbwrappers.cpp#L2689-L2698) function an exception is thrown using the macro [GP_WRAP_END](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gpopt/gpdbwrappers.cpp#L44-L50) and therefore the lines, [testing for NULL](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gpopt/utils/CMemoryPoolPalloc.cpp#L50-L53) are never run and can be safely removed.
And if there is not enough memory for the ORCA's memory pool manager in the [CMemoryPoolTracker::NewImpl](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/memory/CMemoryPoolTracker.cpp#L67-L118) function an exception is thrown using the macro [GPOS_OOM_CHECK](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/memory/CMemoryPoolTracker.cpp#LL90C22-L90C22), but for some reason [before that, the result is checked for NULL and NULL is returned](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/memory/CMemoryPoolTracker.cpp#L85-L88), which can lead to a segfault with further use of the NULL result, so the rows must be deleted.
Unfortunately there is no easy way to test this. We came up with a couple of options, but both of them are not very good: either make a fault-injector, or replace the allocator through preload.

### CODE REVIEW

**Comment by RekGRpth** [comment]
I'll update the commit description if that description is clear.

File: src/backend/gpopt/utils/CMemoryPoolPalloc.cpp [hunk 1]
```diff
@@ -47,11 +47,6 @@ CMemoryPoolPalloc::NewImpl(const ULONG bytes, const CHAR *, const ULONG,
...
 		void *ptr = gpdb::GPDBMemoryContextAlloc(m_cxt, alloc_size);
 
-		if (NULL == ptr)
-		{
-			return NULL;
-		}
-
 		SArrayAllocHeader *header = static_cast<SArrayAllocHeader *>(ptr);
 
 		header->m_user_size = bytes;
```
Comment (HustonMmmavr) [inline src/backend/gpopt/utils/CMemoryPoolPalloc.cpp:52]: Seems here would be a `NULL`-pointer dereference, if ptr is `NULL`. Maybe it's worth to add `GPOS_OOM_CHECK(ptr)` like at `CMemoryPoolTracker`?

Comment (HustonMmmavr) [reply src/backend/gpopt/utils/CMemoryPoolPalloc.cpp:52]: Or `GPDBMemoryContextAlloc` will raise an error? If so, maybe it's worth to add comment about that

Comment (RekGRpth) [reply src/backend/gpopt/utils/CMemoryPoolPalloc.cpp:52]: > Seems here would be a `NULL`-pointer dereference, if ptr is `NULL`.

No! In this case, GPDBMemoryContextAlloc throws an exception! And I mentioned it in the description.

Comment (HustonMmmavr) [reply src/backend/gpopt/utils/CMemoryPoolPalloc.cpp:52]: Sorry, I've forgot to submit review, so this message

> Or GPDBMemoryContextAlloc will raise an error? If so, maybe it's worth to add comment about that

wasn't visible

**Comment by InnerLife0** [comment]
Can you remind me where we left off? There is no easy way of testing this, so we decided to leave this untested?

**Comment by RekGRpth** [comment]
> There is no easy way of testing this

Yes, we came up with a couple of options, but both of them are not very good: either make a fault-injector, or replace the allocator through preload.

**Comment by InnerLife0** [comment]
> > There is no easy way of testing this
> 
> Yes, we came up with a couple of options, but both of them are not very good: either make a fault-injector, or replace the allocator through preload.

Thanks.
Overall looks OK, but we should add a few words about this in description.

Nitpicking on writing, there is some logical inconsistency in existed description. Functions shouldn't return NULL not because of existed exception, but because of their contract, their API. Exception is the way of errors handling and used to confirm our contract.

**Comment by RekGRpth** [comment]
> there is some logical inconsistency in existed description

I've updated the pull request description.

**Comment by InnerLife0** [comment]
> I've updated the pull request description.

Thanks.
What about my first suggestion about adding few words to describe a complexity of testing? If you find some suggestions redundant it's OK to say.

**Review by HustonMmmavr — APPROVED** [review]

**Comment by RekGRpth** [comment]
I've updated the pull request description again.

**Review by InnerLife0 — APPROVED** [review]

**Review by InnerLife0 — APPROVED** [review]

**Review by HustonMmmavr — APPROVED** [review]

**Pull request merged**